libs/gltfio/src/StbProvider.cpp

```diff
@@ -0,0 +1,236 @@
+    ++mPushedCount;
+
+    info->texture = texture;
+    info->state = TextureState::DECODING;
+    info->sourceBuffer.assign(data, data + byteCount);
+    info->decoderJob = nullptr;
+    info->decodedTexelsBaseMipmap.store(0);
+
+    // On single threaded systems, it is usually fine to create jobs because the job system will
+    // simply execute serially. However in our case, we wish to amortize the decoder cost across
+    // several frames, so we instead use the updateQueue() method to perform decoding.
+    if constexpr (!UTILS_HAS_THREADING) {
+        return texture;
+    }
+
+    JobSystem* js = &mEngine->getJobSystem();
+    info->decoderJob = jobs::createJob(*js, mDecoderRootJob, [this, info] {
+        auto& source = info->sourceBuffer;
+        int width, height, comp;
+
+        // Test asynchronous loading by uncommenting this line and adding #include <chrono>.
+        // std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 10000));
+
+        stbi_uc* texels = stbi_load_from_memory(source.data(), source.size(),
+                &width, &height, &comp, 4);
+        source.clear();
+        source.shrink_to_fit();
+        info->decodedTexelsBaseMipmap.store(texels ? intptr_t(texels) : DECODING_ERROR);
+    });
+
+    js->run(info->decoderJob);
+    return texture;
+}
+
+Texture* StbProvider::popTexture() {
+    // We don't bother shrinking the mTextures vector here, instead we periodically clean it up in
+    // the updateQueue method, since popTexture is typically called more frequently.
+    for (auto& texture : mTextures) {
+        if (texture->state == TextureState::READY) {
+            texture->state = TextureState::POPPED;
+            ++mPoppedCount;
+            return texture->texture;
+        }
+    }
+    return nullptr;
+}
+
+void StbProvider::updateQueue() {
+    if (!UTILS_HAS_THREADING) {
+        decodeSingleTexture();
+    }
+    JobSystem* js = &mEngine->getJobSystem();
+    for (auto& info : mTextures) {
+        if (info->state != TextureState::DECODING) {
+            continue;
+        }
+        Texture* texture = info->texture;
+        if (intptr_t data = info->decodedTexelsBaseMipmap.load()) {
+            if (info->decoderJob) {
+                js->waitAndRelease(info->decoderJob);
+                info->decoderJob = nullptr;
+            }
+            if (data == DECODING_ERROR) {
+                slog.e << "stb_image failed to decode texture." << io::endl;
+                info->state = TextureState::READY;
+                ++mDecodedCount;
+                continue;
+            }
+            Texture::PixelBufferDescriptor pbd((uint8_t*) data,
+                    texture->getWidth() * texture->getHeight() * 4, Texture::Format::RGBA,
+                    Texture::Type::UBYTE, [](void* mem, size_t, void*) { stbi_image_free(mem); });
+            texture->setImage(*mEngine, 0, std::move(pbd));
+            texture->generateMipmaps(*mEngine);
```

Comment:
Seems `generateMipmaps` gets called unconditionally. Do we want a way for clients to control this?

Reply:
Good question!  The tl;dr is that I think this is fine, given that glTF is the only user of this interface.  The TextureProvider docstring says that all miplevels should be ready after the decoding signal, and this is done because glTF uses linear mipmap filtering.

The Stb provider can only handle single-image formats so it calls generateMipmaps unconditionally.  The Ktx2 provider will only do this if the ktx file doesn't have all miplevels.